### PR TITLE
Multi-GPU training with DP and DDP documentation

### LINKED ISSUE #488: [Task] Compare models trained with single GPU, DataParallel and DistributedDataParallel with Trainer (NextItemPredictionTask)
### Description
This task is to benchmark the runtime and average examples per second of the following configurations for REES46 for the `NexItemPredictionTask` using the `Trainer`:
- Single GPU
- Multi-GPU with `DataParallel` (2 and 4 GPUs)
- Multi-GPU with `DistributedDataParallel`  (2 and 4 GPUs)

### LINKED ISSUE #492: Documentation on multi-GPU training with DataParallel and DistributedDataParallel
Provide documentation on how to train a Transformers4Rec model using multiple GPU with `DataParallel` (DP) and `DistributedDataParallel` (DDP).

- [x] Short explanation of DP and DDP with links to PyTorch documentation
- [x] Describe code snippets, command line examples and environment variables necessary to use the `Trainer` with `DP` and `DDP`
- [x] Table comparing runtimes x number of GPUs for one of the integration tests for single GPU, `DP` and `DDP`
- [x] Describe that LR needs to be increased when using DP and DDP

### PR DESCRIPTION
**Pull request by nzarif**
Fixes #492 
Fixes #488 
### Goals :soccer:
- Document how to use `DataParallel` and `DistributedDataParallel` multi-gpu training.
- Demonstrate the potential performance benefit of  `DataParallel` and `DistributedDataParallel` multi-gpu training through presenting experiment results.

### Implementation Details :construction:
- Experiments were done on a node with 2 Tesla V100-SXM2-32GB-LS GPUs using T4Rec after changes included in #496. The scripts used were the ones included in [ci/test_integration.sh](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/ci/test_integration.sh).

### Testing Details :mag:
- Refer to the documentation content to use  `DataParallel` or `DistributedDataParallel` mode.

### CODE REVIEW

File: docs/source/multi_gpu_train.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+## Multi-GPU data-parallel training using the `Trainer` class
+
+To train models faster, users can use Data-Parallel training when using `transformers4rec.Trainer` class for training. When using the data-parallel approach for multi-GPU training, the training data will be distributed between workers to speedup training and support larger batch sizes within each step of training.
```
Comment (gabrielspmoreira) [inline docs/source/multi_gpu_train.md:3]:
```suggestion
To train models faster, users can use Data-Parallel training when using `transformers4rec.Trainer` class for training. When using the data-parallel approach for multi-GPU training, the training data will be distributed between the GPUs to speedup training and support larger batch sizes within each step of training.
```

File: docs/source/multi_gpu_train.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+The `Trainer` class supports both `DataParallel` and `DistributedDataParallel` built-in features of PyTorch. in this documentation we will explain how each of these built-in features work and how they can be used.
```
Comment (gabrielspmoreira) [inline docs/source/multi_gpu_train.md:5]:
```suggestion
The `Trainer` class supports both `DataParallel` and `DistributedDataParallel` built-in features of PyTorch. in this documentation we explain how each of these Data Parallel training can be used in Transformers4Rec.
```

File: docs/source/multi_gpu_train.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+### DataParallel
+When the DataParallel mode is used for training on each training step:
```
Comment (gabrielspmoreira) [inline docs/source/multi_gpu_train.md:8]:
```suggestion
When the `DataParallel` mode is used, the following happens for each training step:
```

File: docs/source/multi_gpu_train.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+- GPU0 reads the full batch then evenly distributes it between available GPUs
```
Comment (gabrielspmoreira) [inline docs/source/multi_gpu_train.md:9]:
```suggestion
- GPU-0 reads a batch, that is evenly distributed among the available GPUs
```

File: docs/source/multi_gpu_train.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+- The latest model will be copied to all GPUs
+- A Python thread will be assigned to running `forward()` on each GPU and thepartial loss will be sent to GPU0 to compute total loss
```
Comment (gabrielspmoreira) [inline docs/source/multi_gpu_train.md:11]:
```suggestion
- A Python thread is created for each GPU to run `forward()` step and the partial loss will be sent to GPU-0 to compute the global loss
```

File: docs/source/multi_gpu_train.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+- Computed total loss will be scattered between GPUs and threads will run `backward()`
```
Comment (gabrielspmoreira) [inline docs/source/multi_gpu_train.md:12]:
```suggestion
- Computed global loss is broadcast to all GPUs threads run `backward()`
```

File: docs/source/multi_gpu_train.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+- Gradients from each GPU will be sent to GPU0 and their average will be computed
```
Comment (gabrielspmoreira) [inline docs/source/multi_gpu_train.md:13]:
```suggestion
- Gradients from each GPU are sent to GPU-0 and their average is computed
```

File: docs/source/multi_gpu_train.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+As we see, parallelism in `DataParallel` mode is implemented through Python threads which will be blocked by GIL (Global Interepreter Lock) so `DataParallel` is not the preferred method. Users are advised to use `DistributedDataParallel` instead as it uses multi-processing instead of multi-threading and is better maintained. Also, some models types such as `transfoxl` can not be used with `DataParallel`. To learn more about `DataParalle` refer to [PyTorch documentations](https://pytorch.org/docs/master/generated/torch.nn.DataParallel.html).
```
Comment (gabrielspmoreira) [inline docs/source/multi_gpu_train.md:15]:
```suggestion
As we see, parallelism in `DataParallel` mode is implemented through Python threads which will be blocked by GIL (Global Interepreter Lock) so `DataParallel` is not the preferred method. Users are advised to use `DistributedDataParallel` instead as it uses multi-processing instead of multi-threading and is better maintained. Also, some models types such as `transfoxl` can not be used with `DataParallel`. To learn more about `DataParallel` refer to [PyTorch documentation](https://pytorch.org/docs/master/generated/torch.nn.DataParallel.html).
```

File: docs/source/multi_gpu_train.md [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+-  Run ```export CUDA_VISIBLE_DEVICES=0,1``` in the terminal
+
+Do not try wrapping the model with ```torch.nn.DataParallel``` yourself because that will break automatic wrapping done by `Trainer`.
+<b>Note:</b> When using `DataParallel` the dataloader generates one batch on each train step then the batch will be divided between GPUs so the `per_device_train_batch_size` argument represents the total batch size in this mode, not size of the batch each GPU receives.
+
+### DistributedDataParallel
+
+This is the suggested and more efficient method. When a model is trained using the `DistributedDataParallel` mode:
+- A eparate process will be assigned to each GPU in the beginning and GPU0 will replicate the model on each GPU
```
Comment (gabrielspmoreira) [inline docs/source/multi_gpu_train.md:30]:
```suggestion
- A separate process will be assigned to each GPU in the beginning and GPU-0 will replicate the model on each GPU
```

File: docs/source/multi_gpu_train.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+- On each step each GPU will consume the separate mini-batch produced for it by the dataloader
```
Comment (gabrielspmoreira) [inline docs/source/multi_gpu_train.md:31]:
```suggestion
- On each step each GPU receives a different mini-batch produced by the dataloader
```

File: docs/source/multi_gpu_train.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+- On the backward pass the gradient from GPUs will be averaged for accumulation
+
+To learn more about `DistibutedDataParallel` see the [PyTorch Documentation](https://pytorch.org/docs/master/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel).
```
Comment (gabrielspmoreira) [inline docs/source/multi_gpu_train.md:34]:
```suggestion
To learn more about `DistributedDataParallel` see the [PyTorch Documentation](https://pytorch.org/docs/master/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel).
```

File: docs/source/multi_gpu_train.md [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+To train using the `DistributedDataParallel` mode user should use PyTorch distributed launcher to run the script:
+
+``` python -m torch.distributed.launch --nproc_per_node N_GPU your_script.py --your_arguments```
+
+To have one process per GPU replace `N_GPU` with the number of GPUs you want to use and make sure `CUDA_VISIBLE_DEVICES` is set accordingly.
+
+<b>Note:</b> When using `DistributedDataParallel` number of partitions of the dataset must be equal or larger than number of processes. If a parquet file with small number of partitions is used, try repartitioning it using cudf or pandas before training. The dataloader checks `dataloader.dataset.npartitions` and will repartition if needed but we advise users to repartition the dataset and save it for better efficiency. Use pandas or cudf for repartitioning. Example of repartitioning a parquet file with cudf:
```
Comment (gabrielspmoreira) [inline docs/source/multi_gpu_train.md:42]:
```suggestion
<b>Note:</b> When using `DistributedDataParallel`, our data loader splits data between the GPUs based on dataset partitions. For that reason, the number of partitions of the dataset must be equal or larger than number of processes. If the parquet file has a small number of row groups (partitions) is used, try repartitioning and saving it again with using cudf or pandas before training. The dataloader checks `dataloader.dataset.npartitions` and will repartition if needed but we advise users to repartition the dataset and save it for better efficiency. Use pandas or cudf for repartitioning. Example of repartitioning a parquet file with cudf:
```

File: docs/source/multi_gpu_train.md [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+Choose `row_group_size` such that `nr_rows/row_group_size>=n_proc` because `n_rows=npartition*row_group_size`.
+
+### Performance Comparison
+
+We trained and evaluated a number of models using single GPU, `DataParallel` and `DistributedDataParallel` training mode and the results are shown in the table below. To reproduce, use the models included in [ci/test_integration.sh](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/ci/test_integration.sh).
+
+![Alt text](_images/DP_DDP.png?raw=true "Performance comparison of diffrerent training modes")
+
+These experiments used a node with 2 <i>Tesla V100-SXM2-32GB-LS</i> GPUs.
```
Comment (gabrielspmoreira) [inline docs/source/multi_gpu_train.md:54]:
```suggestion
These experiments used a machine with 2 <i>Tesla V100-SXM2-32GB-LS</i> GPUs.
```

**Review by gabrielspmoreira — APPROVED** [review]

**Comment by gabrielspmoreira** [comment]
@nzarif You need to add a link to your new doc page in the [TOC (Table of Content)](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/docs/source/toc.yaml)

**Review by sararb — APPROVED** [review]
Thank you @nzarif for this detailed documentation, I left two questions. 

Do you think we can add a quick note about the impact of DP/DDP on ndcg/recall metrics and advise the user to tune the learning rate for better accuracy?

**Review by gabrielspmoreira — APPROVED** [review]

File: docs/source/multi_gpu_train.md [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+To train using the `DistributedDataParallel` mode user should use PyTorch distributed launcher to run the script:
+
+``` python -m torch.distributed.launch --nproc_per_node N_GPU your_script.py --your_arguments```
+
+To have one process per GPU replace `N_GPU` with the number of GPUs you want to use and make sure `CUDA_VISIBLE_DEVICES` is set accordingly.
+
+<b>Note:</b> When using `DistributedDataParallel`, our data loader splits data between the GPUs based on dataset partitions. For that reason, the number of partitions of the dataset must be equal or larger than number of processes. If the parquet file has a small number of row groups (partitions), try repartitioning and saving it again using cudf or pandas before training. The dataloader checks `dataloader.dataset.npartitions` and will repartition if needed but we advise users to repartition the dataset and save it for better efficiency. Use pandas or cudf for repartitioning. Example of repartitioning a parquet file with cudf:
+
+```df.to_parquet("filename.parquet", row_group_size=10000, engine="pyarrow"```
```
Comment (gabrielspmoreira) [inline docs/source/multi_gpu_train.md:44]:
```suggestion
```df.to_parquet("filename.parquet", row_group_size=10000)```
```
